File: crates/broomva-cli/src/cli/relay.rs
//! Relay subcommand handlers — remote agent session management.
//!
//! `broomva relay auth`     — registers this machine via device auth
//! `broomva relay start`    — starts life-relayd daemon (exec or inline HTTP polling)
//! `broomva relay stop`     — stops the running daemon
//! `broomva relay status`   — shows nodes, sessions, daemon health
//! `broomva relay sessions` — lists active sessions
//!
//! Auth: reuses the token from `broomva auth login`. The relay daemon
//! reads `~/.broomva/config.json` automatically via life-relayd's
//! token resolution chain.

use std::process::Command as ProcessCommand;

use crate::api::BroomvaClient;
use crate::cli::output::OutputFormat;
use crate::config;
use crate::error::BroomvaResult;

/// Locate the life-relayd binary on PATH or at known locations.
fn find_relayd_binary() -> Option<String> {
    // 1. Check PATH (relayd or life-relayd)
    for name in &["relayd", "life-relayd"] {
        if let Ok(output) = ProcessCommand::new("which").arg(name).output() {
            if output.status.success() {
                let path = String::from_utf8_lossy(&output.stdout).trim().to_string();
                if !path.is_empty() {
                    return Some(path);
                }
            }
        }
    }

    // 2. Check common install locations
    let home = dirs::home_dir().unwrap_or_default();
    let candidates = [
        home.join(".local/bin/relayd"),
        home.join(".local/bin/life-relayd"),
        home.join(".cargo/bin/life-relayd"),
        home.join("broomva/core/life/.target/release/life-relayd"),
    ];

    for path in &candidates {
        if path.exists() {
            return Some(path.to_string_lossy().into_owned());
        }
    }

    None
}

/// Register this machine as a relay node via device auth.
pub async fn handle_auth(client: &BroomvaClient, name: Option<String>) -> BroomvaResult<()> {
    let hostname = hostname::get()
        .map(|h| h.to_string_lossy().into_owned())
        .unwrap_or_else(|_| "unknown".to_string());

    let node_name = name.unwrap_or_else(|| hostname.clone());

    eprintln!("Registering relay node: {node_name}");
    eprintln!("Hostname: {hostname}");
    eprintln!();

    // Use the existing device login flow with relay agent metadata
    let token_response = crate::api::auth::device_login_as_agent(
        client.raw_client(),
        client.base_url(),
        &node_name,
        &hostname,
        &[
            "chat:send",
            "chat:read",
            "deployment:read",
            "deployment:write",
            "memory:read",
            "memory:write",
        ],
    )
    .await?;

    // Store the token
    let computed_expires = token_response.expires_in.map(|secs| {
        let dt = chrono::Utc::now() + chrono::Duration::seconds(secs as i64);
        dt.to_rfc3339()
    });
    let expires_at = token_response
        .expires_at
        .as_deref()
        .or(computed_expires.as_deref());

    config::store_token(&token_response.access_token, expires_at)?;

    eprintln!();
    eprintln!("Relay node registered successfully.");
    eprintln!("Token stored. Run `broomva relay start` to connect.");

    Ok(())
}

/// Start the relay daemon.
///
/// If `life-relayd` binary is found, execs it directly (full PTY + agent support).
/// Otherwise falls back to a lightweight HTTP polling loop.
pub async fn handle_start(_client: &BroomvaClient, bind: &str) -> BroomvaResult<()> {
    // Verify we have a token
    let token = config::resolve_token(None)?;
    if token.is_none() {
        eprintln!("Not authenticated. Run `broomva auth login` first.");
        return Ok(());
    }

    let api_base = config::resolve_api_base(None)?;

    // Try to find and exec the native relay daemon
    if let Some(binary) = find_relayd_binary() {
        eprintln!("Starting relay daemon ({binary})...");
        eprintln!("  Server: {api_base}");
        eprintln!("  Bind:   {bind}");
        eprintln!();

        let status = ProcessCommand::new(&binary)
            .arg("start")
            .arg("--bind")
            .arg(bind)
            .arg("--server")
            .arg(&api_base)
            .env("BROOMVA_TOKEN", token.as_deref().unwrap_or(""))
            .status()
            .map_err(|e| {
                crate::error::BroomvaError::Config(format!("failed to exec {binary}: {e}"))
            })?;

        if !status.success() {
            eprintln!(
                "Relay daemon exited with code {}",
                status.code().unwrap_or(-1)
            );
        }
    } else {
        eprintln!("life-relayd binary not found.");
        eprintln!();
        eprintln!("Installing life-relayd (agent relay daemon)...");

        let install_status = ProcessCommand::new("cargo")
            .arg("install")
            .arg("--git")
            .arg("https://github.com/broomva/life.git")
            .arg("life-relayd")
            .status();

        match install_status {
            Ok(s) if s.success() => {
                eprintln!();
                // Retry with the freshly installed binary
                if let Some(binary) = find_relayd_binary() {
                    eprintln!("Installed. Starting relay daemon ({binary})...");
                    let status = ProcessCommand::new(&binary)
                        .arg("start")
                        .arg("--bind")
                        .arg(bind)
                        .arg("--server")
                        .arg(&api_base)
                        .env("BROOMVA_TOKEN", token.as_deref().unwrap_or(""))
                        .status()
                        .map_err(|e| {
                            crate::error::BroomvaError::Config(format!(
                                "failed to exec {binary}: {e}"
                            ))
                        })?;
                    if !status.success() {
                        eprintln!(
                            "Relay daemon exited with code {}",
                            status.code().unwrap_or(-1)
                        );
                    }
                } else {
                    eprintln!("Install succeeded but binary not found on PATH.");
                    eprintln!(
                        "Try: cargo install --git https://github.com/broomva/life.git life-relayd"
                    );
                }
            }
            _ => {
                eprintln!("Auto-install failed. Falling back to lightweight mode.");
                eprintln!("  To install manually:");
                eprintln!(
                    "    cargo install --git https://github.com/broomva/life.git life-relayd"
                );
                eprintln!();
                run_lightweight_relay(&api_base, &token.unwrap(), bind).await?;
            }
        }
    }

    Ok(())
}

/// Lightweight relay: register node + poll loop (no PTY spawning).
/// Useful for testing the connection or when the Rust daemon isn't installed.
async fn run_lightweight_relay(api_base: &str, token: &str, _bind: &str) -> BroomvaResult<()> {
    let http = reqwest::Client::new();
    let hostname = hostname::get()
        .map(|h| h.to_string_lossy().into_owned())
        .unwrap_or_else(|_| "unknown".to_string());

    // Register node
    eprintln!("Registering node...");
    let connect_res = http
        .post(format!("{api_base}/api/relay/connect"))
        .header("Authorization", format!("Bearer {token}"))
        .json(&serde_json::json!({
            "name": hostname,
            "hostname": hostname,
            "capabilities": ["claude-code"]
        }))
        .send()
        .await
        .map_err(|e| crate::error::BroomvaError::Api {
            status: 0,
            message: format!("connect failed: {e}"),
            body: None,
        })?;

    if !connect_res.status().is_success() {
        let body = connect_res.text().await.unwrap_or_default();
        eprintln!("Node registration failed: {body}");
        return Ok(());
    }

    let connect_body: serde_json::Value = connect_res.json().await.unwrap_or_default();
    let node_id = connect_body
        .get("nodeId")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown");
    let status = connect_body
        .get("status")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown");

    eprintln!("Node {status}: {node_id}");
    eprintln!("Polling for commands (Ctrl+C to stop)...");
    eprintln!();
    eprintln!("NOTE: Lightweight mode — cannot spawn agent sessions.");
    eprintln!("      Install life-relayd for full functionality.");
    eprintln!();

    loop {
        tokio::select! {
            _ = tokio::signal::ctrl_c() => {
                eprintln!();
                eprintln!("Shutting down relay...");
                break;
            }
            _ = tokio::time::sleep(std::time::Duration::from_secs(2)) => {
                let poll_res = http
                    .get(format!("{api_base}/api/relay/poll?nodeId={node_id}"))
                    .header("Authorization", format!("Bearer {token}"))
                    .send()
                    .await;

                match poll_res {
                    Ok(resp) if resp.status().is_success() => {
                        if let Ok(body) = resp.json::<serde_json::Value>().await {
                            if let Some(cmd) = body.get("command") {
                                if !cmd.is_null() {
                                    let cmd_type = cmd.get("type").and_then(|t| t.as_str()).unwrap_or("?");
                                    eprintln!("  Command received: {cmd_type} (cannot execute in lightweight mode)");
                                }
                            }
                        }
                    }
                    Ok(resp) => {
                        eprintln!("  Poll error: HTTP {}", resp.status());
                    }
                    Err(e) => {
                        eprintln!("  Poll error: {e}");
                    }
                }
            }
        }
    }

    Ok(())
}

/// Stop the relay daemon.
pub async fn handle_stop() -> BroomvaResult<()> {
    let client = reqwest::Client::new();
    match client
        .get("http://127.0.0.1:3004/health")
        .timeout(std::time::Duration::from_secs(2))
        .send()
        .await
    {
        Ok(resp) if resp.status().is_success() => {
            eprintln!("Relay daemon is running on port 3004.");
            eprintln!("Send Ctrl+C to the running process, or:");
            eprintln!("  kill $(lsof -ti :3004)");
        }
        _ => {
            eprintln!("No relay daemon running on port 3004.");
        }
    }
    Ok(())
}

/// Show relay status — nodes, sessions, daemon health.
pub async fn handle_status(client: &BroomvaClient, _format: OutputFormat) -> BroomvaResult<()> {
    // Check local daemon health
    let reqw = reqwest::Client::new();
    let daemon_running = reqw
        .get("http://127.0.0.1:3004/health")
        .timeout(std::time::Duration::from_secs(2))
        .send()
        .await
        .is_ok_and(|r| r.status().is_success());

    // Check remote nodes
    let url = format!("{}/api/relay/nodes", client.base_url());
    let mut req = client.raw_client().get(&url);
    if let Ok(Some(t)) = config::read_config().map(|c| c.token) {
        req = req.header("Authorization", format!("Bearer {t}"));
    }
    let res = req
        .send()
        .await
        .map_err(|e| crate::error::BroomvaError::Api {
            status: 0,
            message: format!("request failed: {e}"),
            body: None,
        })?;

    let body: serde_json::Value = res.json().await.unwrap_or_default();

    eprintln!();
    eprintln!("  Relay Status");
    eprintln!("  ─────────────────────────────");
    eprintln!(
        "  Authenticated:  {}",
        if config::resolve_token(None)?.is_some() {
            "yes"
        } else {
            "no"
        }
    );
    eprintln!(
        "  Local daemon:   {}",
        if daemon_running {
            "running (port 3004)"
        } else {
            "not running"
        }
    );
    eprintln!(
        "  Relayd binary:  {}",
        find_relayd_binary().as_deref().unwrap_or("not found")
    );

    if let Some(nodes) = body.get("nodes").and_then(|n| n.as_array()) {
        eprintln!();
        eprintln!("  Nodes ({}):", nodes.len());
        for node in nodes {
            let name = node.get("name").and_then(|v| v.as_str()).unwrap_or("?");
            let status = node.get("status").and_then(|v| v.as_str()).unwrap_or("?");
            let host = node.get("hostname").and_then(|v| v.as_str()).unwrap_or("");
            let indicator = match status {
                "online" => "●",
                "degraded" => "◐",
                _ => "○",
            };
            eprintln!("    {indicator} {name} ({host}) — {status}");
        }
    }

    if let Some(metrics) = body.get("metrics") {
        let online = metrics
            .get("nodesOnline")
            .and_then(|n| n.as_u64())
            .unwrap_or(0);
        let active = metrics
            .get("sessionsActive")
            .and_then(|n| n.as_u64())
            .unwrap_or(0);
        eprintln!();
        eprintln!("  {online} node(s) online, {active} active session(s)");
    }

    if !daemon_running {
        eprintln!();
        eprintln!("  Run `broomva relay start` to connect this machine.");
    }

    eprintln!();
    Ok(())
}

/// List relay sessions.
pub async fn handle_sessions(client: &BroomvaClient, _format: OutputFormat) -> BroomvaResult<()> {
    let url = format!("{}/api/relay/sessions", client.base_url());
    let mut req = client.raw_client().get(&url);
    if let Ok(Some(t)) = config::read_config().map(|c| c.token) {
        req = req.header("Authorization", format!("Bearer {t}"));
    }
    let res = req
        .send()
        .await
        .map_err(|e| crate::error::BroomvaError::Api {
            status: 0,
            message: format!("request failed: {e}"),
            body: None,
        })?;

    let body: serde_json::Value = res.json().await.unwrap_or_default();

    if let Some(sessions) = body.get("sessions").and_then(|s| s.as_array()) {
        if sessions.is_empty() {
            eprintln!("No active relay sessions.");
            return Ok(());
        }
        eprintln!("Relay Sessions:");
        for session in sessions {
            let name = session
                .get("name")
                .and_then(|v| v.as_str())
                .unwrap_or("untitled");
            let stype = session
                .get("sessionType")
                .and_then(|v| v.as_str())
                .unwrap_or("?");
            let status = session
                .get("status")
                .and_then(|v| v.as_str())
                .unwrap_or("?");
            let workdir = session
                .get("workdir")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            eprintln!("  [{status:9}] [{stype:11}] {name}  {workdir}");
        }
    }

    Ok(())
}

File: packages/cli/src/commands/relay.ts
/**
 * `broomva relay` — manage the relay daemon for remote agent sessions.
 *
 * Connects your local machine to broomva.tech so you can run Claude Code,
 * Codex, or Arcan sessions from the web console.
 *
 * Auth: reuses the token from `broomva auth login`. If already authenticated,
 * `broomva relay start` works immediately — no separate auth step.
 */

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { Command } from "commander";
import { resolveToken } from "../lib/auth-store.js";
import { DEFAULT_API_BASE } from "../lib/constants.js";
import { fmt, info, error as printError, success, warn } from "../lib/output.js";

// ─── Config ──────────────────────────────────────────────────────────────────

const RELAY_CONFIG_DIR = join(homedir(), ".broomva", "relay");
const RELAY_CREDENTIALS = join(RELAY_CONFIG_DIR, "credentials.json");
const RELAY_PID_FILE = join(RELAY_CONFIG_DIR, "relay.pid");
const DEFAULT_RELAY_PORT = 3004;
const DEFAULT_POLL_INTERVAL = 2000;

function ensureRelayDir(): void {
	mkdirSync(RELAY_CONFIG_DIR, { recursive: true });
}

/**
 * Resolve relay auth token. Priority:
 * 1. Relay-specific credentials (~/.broomva/relay/credentials.json)
 * 2. Broomva CLI token (~/.broomva/config.json via resolveToken)
 * 3. Life relay credentials (~/.config/life/relay/credentials.json)
 * 4. BROOMVA_TOKEN env var
 */
function resolveRelayToken(): { token: string; source: string } | null {
	// 1. Relay-specific credentials
	if (existsSync(RELAY_CREDENTIALS)) {
		try {
			const data = JSON.parse(readFileSync(RELAY_CREDENTIALS, "utf-8"));
			if (data.token) return { token: data.token, source: "relay credentials" };
		} catch {}
	}

	// 2. Broomva CLI token
	const cliToken = resolveToken();
	if (cliToken) return { token: cliToken.token, source: `broomva CLI (${cliToken.source})` };

	// 3. Life relay credentials (Rust daemon path)
	const lifeRelayPaths = [
		join(homedir(), "Library", "Application Support", "life", "relay", "credentials.json"),
		join(homedir(), ".config", "life", "relay", "credentials.json"),
	];
	for (const p of lifeRelayPaths) {
		if (existsSync(p)) {
			try {
				const data = JSON.parse(readFileSync(p, "utf-8"));
				if (data.token) return { token: data.token, source: "life relay credentials" };
			} catch {}
		}
	}

	// 4. Env var
	if (process.env.BROOMVA_TOKEN) {
		return { token: process.env.BROOMVA_TOKEN, source: "BROOMVA_TOKEN env" };
	}

	return null;
}

// ─── Commands ────────────────────────────────────────────────────────────────

export function relayCommand(): Command {
	const relay = new Command("relay").description(
		"Manage the relay daemon — connect your machine to broomva.tech for remote agent sessions",
	);

	// ── relay auth ──
	relay
		.command("auth")
		.description("Authenticate relay with broomva.tech (reuses broomva auth login if available)")
		.option("--url <url>", "Server URL", DEFAULT_API_BASE)
		.action(async (opts: { url: string }) => {
			const existing = resolveRelayToken();
			if (existing) {
				success(`Already authenticated (${existing.source})`);
				console.log();
				console.log(`  Token source: ${fmt.dim(existing.source)}`);
				console.log(`  To re-authenticate: ${fmt.cyan("broomva auth login")}`);
				console.log();
				return;
			}

			// No token found — guide user to broomva auth login
			warn("No authentication token found.");
			console.log();
			console.log(`  Run ${fmt.bold("broomva auth login")} to authenticate.`);
			console.log(`  The relay will automatically use your broomva.tech session.`);
			console.log();
		});

	// ── relay start ──
	relay
		.command("start")
		.description("Start the relay daemon (connects to broomva.tech, polls for commands)")
		.option("--server <url>", "Server URL", DEFAULT_API_BASE)
		.option("--port <port>", "Local API port", String(DEFAULT_RELAY_PORT))
		.option("--name <name>", "Node display name")
		.action(async (opts: { server: string; port: string; name?: string }) => {
			const auth = resolveRelayToken();
			if (!auth) {
				printError("Not authenticated. Run `broomva auth login` first.");
				process.exit(1);
			}

			const port = Number.parseInt(opts.port, 10);
			const hostname = opts.name || require("node:os").hostname();

			info(`Starting relay daemon...`);
			console.log(`  Server:   ${fmt.cyan(opts.server)}`);
			console.log(`  Node:     ${fmt.bold(hostname)}`);
			console.log(`  API port: ${port}`);
			console.log(`  Auth:     ${fmt.dim(auth.source)}`);
			console.log();

			ensureRelayDir();

			// Register node
			info("Registering node...");
			try {
				const connectRes = await fetch(`${opts.server}/api/relay/connect`, {
					method: "POST",
					headers: {
						"Content-Type": "application/json",
						Authorization: `Bearer ${auth.token}`,
					},
					body: JSON.stringify({
						name: hostname,
						hostname,
						capabilities: ["claude-code", "codex", "arcan"],
					}),
				});

				if (!connectRes.ok) {
					const err = await connectRes.json().catch(() => ({ error: "Unknown error" }));
					printError(`Node registration failed: ${err.error || connectRes.statusText}`);
					process.exit(1);
				}

				const { nodeId, status } = (await connectRes.json()) as {
					nodeId: string;
					status: string;
				};
				success(`Node ${fmt.bold(status)}: ${fmt.dim(nodeId)}`);

				// Write PID
				writeFileSync(RELAY_PID_FILE, String(process.pid));

				// Start polling loop
				info("Starting command polling loop (Ctrl+C to stop)...");
				console.log();

				// Graceful shutdown
				const shutdown = () => {
					console.log();
					info("Shutting down relay daemon...");
					try {
						if (existsSync(RELAY_PID_FILE))
							require("node:fs").unlinkSync(RELAY_PID_FILE);
					} catch {}
					process.exit(0);
				};
				process.on("SIGINT", shutdown);
				process.on("SIGTERM", shutdown);

				// Poll loop
				let consecutiveErrors = 0;
				while (true) {
					try {
						const pollRes = await fetch(
							`${opts.server}/api/relay/poll?nodeId=${nodeId}`,
							{
								headers: { Authorization: `Bearer ${auth.token}` },
							},
						);

						if (pollRes.ok) {
							consecutiveErrors = 0;
							const { command } = (await pollRes.json()) as {
								command: Record<string, unknown> | null;
							};
							if (command) {
								console.log(
									`  ${fmt.green("→")} Command: ${fmt.bold(String(command.type))} ${fmt.dim(JSON.stringify(command).slice(0, 80))}`,
								);
							}
						} else {
							consecutiveErrors++;
							if (consecutiveErrors > 10) {
								printError("Too many consecutive poll errors. Stopping.");
								process.exit(1);
							}
						}
					} catch (err) {
						consecutiveErrors++;
						if (consecutiveErrors > 10) {
							printError(
								`Connection lost: ${err instanceof Error ? err.message : "unknown"}`,
							);
							process.exit(1);
						}
					}

					await new Promise((r) => setTimeout(r, DEFAULT_POLL_INTERVAL));
				}
			} catch (err) {
				printError(
					`Failed to start: ${err instanceof Error ? err.message : "unknown error"}`,
				);
				process.exit(1);
			}
		});

	// ── relay stop ──
	relay
		.command("stop")
		.description("Stop the relay daemon")
		.action(() => {
			if (!existsSync(RELAY_PID_FILE)) {
				warn("No relay daemon running.");
				return;
			}

			try {
				const pid = Number.parseInt(
					readFileSync(RELAY_PID_FILE, "utf-8").trim(),
					10,
				);
				process.kill(pid, "SIGTERM");
				success(`Sent stop signal to relay daemon (PID ${pid})`);
			} catch {
				warn("Could not stop daemon (may have already exited).");
			}

			try {
				require("node:fs").unlinkSync(RELAY_PID_FILE);
			} catch {}
		});

	// ── relay status ──
	relay
		.command("status")
		.description("Show relay daemon and connection status")
		.option("--json", "Output as JSON")
		.action(async (opts: { json?: boolean }) => {
			const auth = resolveRelayToken();
			const daemonRunning = existsSync(RELAY_PID_FILE);
			let daemonPid: number | null = null;

			if (daemonRunning) {
				try {
					daemonPid = Number.parseInt(
						readFileSync(RELAY_PID_FILE, "utf-8").trim(),
						10,
					);
					process.kill(daemonPid, 0); // check if alive
				} catch {
					daemonPid = null;
				}
			}

			// Check local API health
			let localApiOk = false;
			let version = "";
			try {
				const res = await fetch(`http://127.0.0.1:${DEFAULT_RELAY_PORT}/health`, {
					signal: AbortSignal.timeout(2000),
				});
				if (res.ok) {
					localApiOk = true;
					const body = (await res.json()) as { version?: string };
					version = body.version || "";
				}
			} catch {}

			if (opts.json) {
				console.log(
					JSON.stringify({
						authenticated: !!auth,
						authSource: auth?.source ?? null,
						daemonRunning: !!daemonPid || localApiOk,
						daemonPid,
						localApiPort: DEFAULT_RELAY_PORT,
						version: version || null,
					}),
				);
				return;
			}

			console.log();
			console.log(`  ${fmt.bold("Relay Status")}`);
			console.log(`  ${"─".repeat(30)}`);
			console.log(
				`  Authenticated:  ${auth ? fmt.green("yes") : fmt.red("no")}${auth ? fmt.dim(` (${auth.source})`) : ""}`,
			);
			console.log(
				`  Daemon:         ${daemonPid || localApiOk ? fmt.green("running") : fmt.dim("not running")}${daemonPid ? fmt.dim(` (PID ${daemonPid})`) : ""}`,
			);
			if (version)
				console.log(`  Version:        ${version}`);
			console.log(
				`  Local API:      ${localApiOk ? fmt.green(`http://127.0.0.1:${DEFAULT_RELAY_PORT}`) : fmt.dim("unavailable")}`,
			);
			console.log();

			if (!auth) {
				console.log(
					`  Run ${fmt.bold("broomva auth login")} to authenticate.`,
				);
				console.log();
			} else if (!daemonPid && !localApiOk) {
				console.log(
					`  Run ${fmt.bold("broomva relay start")} to connect this machine.`,
				);
				console.log();
			}
		});

	// ── relay setup (guided) ──
	relay
		.command("setup")
		.description("Guided setup: authenticate and start the relay daemon")
		.action(async () => {
			console.log();
			console.log(`  ${fmt.bold("Relay Setup")}`);
			console.log(`  ${"─".repeat(30)}`);
			console.log();

			// Step 1: Check auth
			const auth = resolveRelayToken();
			if (auth) {
				success(`Step 1: Authenticated (${auth.source})`);
			} else {
				warn("Step 1: Not authenticated");
				console.log(`  Run ${fmt.bold("broomva auth login")} first, then re-run this command.`);
				console.log();
				process.exit(1);
			}

			// Step 2: Check if daemon already running
			let localApiOk = false;
			try {
				const res = await fetch(`http://127.0.0.1:${DEFAULT_RELAY_PORT}/health`, {
					signal: AbortSignal.timeout(2000),
				});
				localApiOk = res.ok;
			} catch {}

			if (localApiOk) {
				success("Step 2: Relay daemon already running");
				console.log();
				console.log(`  Your machine is connected to broomva.tech.`);
				console.log(`  Open ${fmt.cyan("https://broomva.tech/console/relay")} to start a session.`);
				console.log();
				return;
			}

			// Step 3: Guide to start
			info("Step 2: Start the relay daemon");
			console.log();
			console.log(`  Run: ${fmt.bold("broomva relay start")}`);
			console.log();
			console.log(`  This connects your machine to broomva.tech so you can`);
			console.log(`  run Claude Code, Codex, or Arcan sessions from the web console.`);
			console.log();
		});

	return relay;
}
